src/main/java/com/hillel/items_exchange/service/impl/CategoryServiceImpl.java

```diff
@@ -0,0 +1,115 @@
+                CategoryDto.class, ArrayList::new));
+    }
+
+    @Override
+    public Optional<CategoryDto> findCategoryDtoById(long id) {
+        return categoryRepository.findById(id)
+                .map(category -> convertTo(category, CategoryDto.class));
+    }
+
+    @Override
+    public CategoryDto create(CategoryDto categoryDto) {
+        final Category savedCategory = saveCategory(categoryDto);
+        return convertTo(savedCategory, CategoryDto.class);
+    }
+
+    @Override
+    public CategoryDto update(CategoryDto categoryDto) {
+        final Category updatedCategory = saveCategory(categoryDto);
+        return convertTo(updatedCategory, CategoryDto.class);
+    }
+
+    @Override
+    public void removeById(long categoryId) {
+        categoryRepository.deleteById(categoryId);
+    }
+
+    @Override
+    public boolean isCategoryDtoDeletable(long categoryId) {
+        return isCategoryExistsById(categoryId) && isInternalSubcategoriesHaveNotProducts(categoryId);
+    }
+
+    @Override
+    public boolean isCategoryDtoUpdatable(CategoryDto categoryDto) {
+        return isCategoryExistsByIdAndNameOrNotExistsByName(categoryDto.getId(),
+                categoryDto.getName()) && isSubcategoryHasIdZeroOrIdExists(categoryDto);
+    }
+
+    @Override
+    public boolean isCategoryDtoValidForCreating(CategoryDto categoryDto) {
+        return isCategoryNameHasNotDuplicate(categoryDto.getName())
+                && categoryDto.getSubcategories().stream().allMatch(subcategoryDto -> subcategoryDto.getId() == 0L);
+    }
+
+    private boolean isCategoryExistsById(long categoryId) {
+        return categoryRepository.existsById(categoryId);
+    }
+
+    private boolean isCategoryNameHasNotDuplicate(String name) {
+        return !categoryRepository.existsByNameIgnoreCase(name);
+    }
+
+    private boolean isSubcategoryHasIdZeroOrIdExists(CategoryDto categoryDto) {
+        return categoryDto.getSubcategories().stream()
+                .filter(subcategoryDto -> subcategoryDto.getId() != 0)
+                .allMatch(subcategoryDto -> subcategoryService.isSubcategoryExistsById(subcategoryDto.getId()));
+    }
+
+    private boolean isInternalSubcategoriesHaveNotProducts(long categoryId) {
+        return categoryRepository.findById(categoryId)
+                .map(category -> category.getSubcategories().stream()
+                        .allMatch(subcategory -> subcategory.getProducts().isEmpty()))
+                .orElse(false);
+    }
+
+    private boolean isCategoryExistsByIdAndNameOrNotExistsByName(long categoryId, String categoryName) {
+        return categoryRepository.existsByIdAndNameIgnoreCase(categoryId, categoryName)
+                || (categoryRepository.existsById(categoryId) && isCategoryNameHasNotDuplicate(categoryName));
+    }
+
+    private Category saveCategory(CategoryDto categoryDto) {
+        final Category category = convertTo(categoryDto, Category.class);
```

Comment:
Have you tried to convert CategoryDto including List<SubcategoryDto> to Category without invoking additional methods?

Reply:
Yes, but it didn't work